prov/tcp/src/tcpx_eq.c

```diff
@@ -61,6 +61,28 @@ static struct fi_ops_eq tcpx_eq_ops = {
 	.strerror = ofi_eq_strerror,
 };
 
+int tcpx_eq_match_ep_event(struct slist_entry *entry, const void *arg)
+{
+	struct tcpx_ep *ep = (struct tcpx_ep *) arg;
+	struct util_event *event;
+	fid_t fid;
+
+	event = container_of(entry, struct util_event, entry);
+
+		fid = (event->err)?
+			((struct fi_eq_err_entry *)event->data)->fid:
+			((struct fi_eq_cm_entry *)event->data)->fid;
+
+	return (&ep->util_ep.ep_fid.fid == fid);
+}
+
+void tcpx_eq_remove_ep_events(struct util_eq *eq, struct tcpx_ep *ep)
+{
+	fastlock_acquire(&eq->lock);
+	while(slist_remove_first_match(&eq->list, tcpx_eq_match_ep_event, ep));
+	fastlock_release(&eq->lock);
+}
```

Comment:
The above two functions could be made more generic by simply taking as input a fid

Reply:
the match function assumes that entries are of  type _fi_eq_err_entry_ or _fi_eq_cm_entry_ . So can that be made common function as well?

Reply:
Thinking about this more, we can probably add this as part of the util eq code.  You might need 2 calls -- not sure.  For the first call, if you take as input a fid, you can remove all eq entries that reference that fid.  Just cast the eq entry to struct fi_eq_entry.

For cm events, you can define a separate call that only matches on cm events and removes any where the data field matches, regardless of what data actually represents.

Does this make sense?

Reply:
why do we need to remove cm events individually?

Reply:
Yeah, I wasn't clear, and was off some.

There are 3 eq structures defined -- fi_eq_entry, fi_eq_err_entry, and fi_eq_cm_entry.  All of these contain a fid pointer as the first element.  A generic function can walk the EQ and remove any entry where that fid matches an input fid.

Additionally, the fi_eq_cm_entry contains an fi_info pointer.  As part of cleanup, we also need to remove any cm entries where fi_info::fid matches.

So a single function can do the work, but cm events need a little more work.  Otherwise we could just cast the eq entries to struct fi_eq_entry and check the fid.

Reply:
Got it now. Thanks